Write unit test for 'enhanced_summary'

d-geula opened the pull request:
* Separated the category extractor code from the test and put it in `tests\utils`.
* Finished writing the test for `enhanced_summary`.
* ~~Renamed and moved the langchain cache for simplicity, and created a new cache db for test functions.~~ 
* Switched back to using just one cache DB for everything. I think it anyway only goes to one, even if there are multiple DBs.
* Added a fake OpenAI API key to the GitHub Actions workflow as an environment variable. Since the test uses the cached results and doesn't need a real key, I did this to bypass the input validation in the `ChatOpenAI` class which was causing the test to fail without it.

d-geula committed:
Create test_enhanced_summary.py

d-geula committed:
Rename langchain cache and create test cache

d-geula committed:
Create summary_category_extractor.py

d-geula committed:
Merge branch 'main' into test-enhanced_summary

d-geula committed:
Rename and move cache db

d-geula committed:
Update test_enhanced_summary.py

d-geula committed:
Update test_cache.db

d-geula committed:
Rebuild cache

d-geula committed:
Update requirements.txt

d-geula committed:
Merge branch 'main' into test-enhanced_summary

itaifrenkel commented on tests/nl2sql/test_enhanced_summary.py:
```diff
@@ -0,0 +1,25 @@
+        sql_results=[('Martin', 'Brundle', '1959-06-01', 'British')],
+    )
+
+    categories = ast.literal_eval(summary_category_extractor(
```
consider "hiding" the dictionary evaluation inside the test utility, so it is easier to use.

itaifrenkel commented on tests/nl2sql/test_enhanced_summary.py:
```diff
@@ -0,0 +1,25 @@
+
+    categories = ast.literal_eval(summary_category_extractor(
+        summary=summary,
+        question_1="Who is this summary about?",
```
```
assert_summary_using_qa(
summary=summary, 
questions_and_answers={
"Who is this summary about?": "Martin Brundle",  # list is now optional 
"Did they win a formula 1 championship?": ['No']
})
```

itaifrenkel commented on tests/utils/summary_category_extractor.py:
```diff
@@ -0,0 +1,48 @@
+
+    human_template = """
+    summary: {summary}
+    qestions: ["{q1}", "{q2}"]"""
```
you got a typo here
and also support more than two questions.

itaifrenkel commented on tests/nl2sql/test_enhanced_summary.py:
```diff
@@ -0,0 +1,25 @@
+    summary = enhanced_summary(
+        docs=Path("docs/martin_brundle.txt").read_text(),
+        user_query="Tell me about Martin Brundle",
+        sql_results=[('Martin', 'Brundle', '1959-06-01', 'British')],
```
```
summary = expand_on_results(
        docs=[Path("docs/martin_brundle.txt")] # .read_text() inside the expand_on_results implementation
        user_query="Tell me about Martin Brundle",
        sql_results={'First Name':'Martin', 'Last Name':'Brundle', 'Date of Birth':'1959-06-01', 'Nationality':'British'}
        )
        ```

itaifrenkel commented on tests/utils/summary_category_extractor.py:
```diff
@@ -0,0 +1,48 @@
+)
+
+
+def summary_category_extractor(summary: str, question_1: str, question_2: str):
```
rename to assert_summary_using_qa (summary, questions)
and internally perform the assertions.

d-geula committed:
Address comments

d-geula commented:
Ok, I think I've addressed most of the comments. 
* `assert_summary_using_qa`: I'm not sure about the way I'm doing the asserts here. I tweaked the system prompt so that it should basically always return single values per question, but I left in the code I wrote to un-list (I don't know how to say it) a single value list just in case it randomly decides to generate it that way. I guess the prompt change does mean that I have to ask it questions with simple, one-value answers, but I also feel like it makes testing a lot more straightforward so maybe it's good.
* `test_expand_on_results`: I changed the questions a bit. I did it partly because of the changes I made to the prompt, but also because for some reason the LLM decided that actually Martin Brundle has won an F1 championship, and because I couldn't figure out why it changed its usual output, I didn't want to change the answer to something incorrect just to pass the test. I also added a third question.
* `expand_on_results`: I didn't manage to add the column names to the SQL result as you suggested. If I were to use the existing function I wrote it wouldn't even do anything for this particular query because it selects everything (SELECT *) which means I can't get all the column names it actually used. I don't know if you had something else in mind, but tbh the only thing I can think of that would solve this without a bunch of hacky solutions is to write my own SQL chain. I think I could do it pretty quickly but I didn't want to just do it without asking because it would still take some time and we didn't plan/agree to do it. At least it would be easier to use/modify, and I wouldn't have to bother with parsing SQL queries anymore, even if it was less sophisticated in some ways than the langchain version.

itaifrenkel commented on nl2sql/expand_on_results.py:
```diff
@@ -32,6 +35,8 @@ def enhanced_summary(docs: str, user_query: str, sql_results: list):
     return chat(
         chat_prompt.format_prompt(
-            docs=docs, user_query=user_query, sql_results=sql_results
+            docs=[Path(doc).read_text() for doc in docs],
```
It looks like format_prompt is just a simple python string formatting:
https://github.com/langchain-ai/langchain/blob/0dd2c21089d82445ef9ee8dd026be1f3194ed4fe/libs/langchain/langchain/prompts/base.py#L117

So it would format docs as an array instead of plain text.
Try these commands in a python interpreter:
```
print("{docs}".format(docs=['one','two','three']))
print("{docs}".format(docs='\n'.join(['one','two','three'])))
```

itaifrenkel replied:
I recommend that you log the final prompt after formatting and look for any quirks with docs or the sql_results.

itaifrenkel commented on tests/utils/assert_summary_using_qa.py:
```diff
@@ -0,0 +1,55 @@
+    )
+
+    # Convert the chat result to a dictionary
+    result = ast.literal_eval(chat(
```
There is too much going on in this single command.
perhaps move the parsing to a separate line ?
```
chat_response=chat(...)
actual_answers=ast.literal_eval(chat_response.content)
```

itaifrenkel commented on tests/utils/assert_summary_using_qa.py:
```diff
@@ -0,0 +1,55 @@
+
+    for question, expected_answer in questions_and_answers.items():
+        answer = result[question]
+        if isinstance(answer, list) and len(answer) == 1:
```
if its a list with length other than 1, you should raise a ValueError() with the answer and a message saying you were expected a single string answer.

itaifrenkel commented:
@d-geula I added more code comments. Let's make another round of fixes and make it first priority focusing on merging this PR.

Re: `test_expand_on_results` - *After* this PR is merged you should add back the test question about winning the formula-1 race and manually try editing different prompts to get to the bottom of this. It could be any number of things, but debugging prompt engineering is a core part of working with LLMs, and we should invest more time in it. Let's make this our second priority. You can do time capping of a few hours, and report back the results.

Re: `expand_on_results` - I understand your point. Is it possible to change the sql chain to a custom prompt such that the result is a query with column names?

d-geula committed:
.

d-geula commented:
I may have overstepped a bit with some of the changes I made tonight so I'm happy to revert anything you're unhappy with or just think should be in a separate PR.

* **`sql_chain`**: starting with the biggest change, I made a new utility function that executes the generated query and returns a dataframe with all the columns and values. I know this is just doing the same thing twice, but from what I can gather it's not the prompt that's the problem when it comes to the issue of column names, it's how langchain is using SQLAlchemy. I also removed the `answer` variable because I wasn't even using it anymore.

* **`app`**:  since now `sql_results` is a dataframe, I don't need to convert it to one using `extract_columns`. I'm also passing the df as a dict to `expand_on_results` so it looks like your previous suggestion (other than it's a list of dicts).
* **`expand_on_results`**: used a `join` on the doc, as well as now having the proper results (including columns) be used in the prompt. I also added `dedent` to the prompts because I realised that there were tons of pointless spaces being sent to the LLM.
* **`assert_summary_using_qa`**: again, added `dedent` for the same reason. Also, if you don't I removed the `isinstance` bit for now because I not sure exactly what you meant and I'm also a bit worried that it was based on my poor explanation of why I added it in the first place. Anyway, it's kind of pointless since I changed the example in the system prompt because it doesn't return anything in the format anymore.
* **`utils`**: got rid of `extract_columns` as it's now unnecessary. Again, I don't mind reverting or moving big changes like this if you want.

As for tests, other than removing the tests for `extract_columns`, I definitely need some feedback on how I updated `test_sql_chain`. Because I'm now returning a dataframe from `sql_chain` I couldn't keep the existing results assert. TBH I didn't write that code myself but from what I can understand the test seems to be actually doing something and it does pass. I could be very wrong about that though so do let me know.

itaifrenkel commented on nl2sql/expand_on_results.py:
```diff
@@ -0,0 +1,43 @@
+                               SystemMessagePromptTemplate)
+
+
+def expand_on_results(docs: Union[str, list], user_query: str, sql_results: list[dict]) -> str:
```
I think it would be cleaner if `sql_result` type was a `pd.DataFrame`. It is more common than `list[dict]` especially when passing tabular data.

itaifrenkel commented on nl2sql/sql_chain.py:
```diff
@@ -15,10 +15,9 @@ def sql_chain(input: str):
     sql_query = result["intermediate_steps"][1]
-    sql_results = ast.literal_eval(result["intermediate_steps"][3])
-    answer = result["intermediate_steps"][5]
+    sql_results = execute_query_and_return_df(sql_query, db_path)
```
Consider naming the method `execute_query`. it is uncommon to add the return type to the method name.

itaifrenkel replied:
Consider extracting the following code to a separate function, so it is clear that it's only purpose is to translate a natural language into an sql query.
```
db = SQLDatabase.from_uri(f"sqlite:///{db_path}", sample_rows_in_table_info=3)
db_chain = SQLDatabaseChain.from_llm(
        _create_llm(), db, verbose=True, return_intermediate_steps=True
    )
result = db_chain(input)
sql_query = result["intermediate_steps"][1]
```

itaifrenkel commented on tests/nl2sql/test_sql_chain.py:
```diff
@@ -17,20 +19,15 @@ def test_sql_chain(llm):
+            "forename": ["Stefan", "Slim", "Ronnie", "Gunnar", "Conny"],
+            "surname": ["Johansson", "Borgudd", "Peterson", "Nilsson", "Andersson"]
+        })
+        assert_frame_equal(sql_results, expected_results)
```
this is really cool. 
You are assuming you know the order of the names returned by the database. You can add `sort_values(["forename", "surname"])`) to both data frames, so even if the order changes the test still passes.

d-geula committed:
.

d-geula commented:
The only thing I think is worth mentioning is that as a result of moving `to_dict` into `expand_on_results` I changed `sql_results` in `expand_on_results` to be a dataframe. I did this mostly because after changing the type as you suggested, vs code was complaining about a type mismatch because I was actually passing a `list[dict]` to `expand_on_results` (which is why I used it originally) since it was being converted in `app.py`.

As for splitting `sql_chain`, do you mind if we go over it briefly next time we talk because it would involve renaming and possibly moving things around a bit so I'd rather do it with you so it looks nice.
